File: algorithms/Java/queues/circular-queue-linked-list.java
// Java program for insertion and
// deletion in Circular Queue
import java.util.*;
  
class Solution {
  
    // Structure of a Node
    static class Node {
        int data;
        Node link;
    }
  
    static class Queue {
        Node front, rear;
    }
  
    // Function to create Circular queue
    static void enQueue(Queue q, int value)
    {
        Node temp = new Node();
        temp.data = value;
        if (q.front == null)
            q.front = temp;
        else
            q.rear.link = temp;
  
        q.rear = temp;
        q.rear.link = q.front;
    }
  
    // Function to delete element from Circular Queue
    static int deQueue(Queue q)
    {
        if (q.front == null) {
            System.out.printf("Queue is empty");
            return Integer.MIN_VALUE;
        }
  
        // If this is the last node to be deleted
        int value; // Value to be dequeued
        if (q.front == q.rear) {
            value = q.front.data;
            q.front = null;
            q.rear = null;
        }
        else // There are more than one nodes
        {
            Node temp = q.front;
            value = temp.data;
            q.front = q.front.link;
            q.rear.link = q.front;
        }
  
        return value;
    }
  
    // Function displaying the elements of Circular Queue
    static void displayQueue(Queue q)
    {
        Node temp = q.front;
        System.out.printf("\nElements in Circular Queue are: ");
        while (temp.link != q.front) {
            System.out.printf("%d ", temp.data);
            temp = temp.link;
        }
        System.out.printf("%d", temp.data);
    }
  
    /* Driver of the program */
    public static void main(String args[])
    {
        // Create a queue and initialize front and rear
        Queue q = new Queue();
        q.front = q.rear = null;
  
        // Inserting elements in Circular Queue
        enQueue(q, 14);
        enQueue(q, 22);
        enQueue(q, 6);
  
        // Display elements present in Circular Queue
        displayQueue(q);
  
        // Deleting elements from Circular Queue
        System.out.printf("\nDeleted value = %d", deQueue(q));
        System.out.printf("\nDeleted value = %d", deQueue(q));
  
        // Remaining elements in Circular Queue
        displayQueue(q);
  
        enQueue(q, 9);
        enQueue(q, 20);
        displayQueue(q);
    }
}
  

File: algorithms/C++/Queues/circular-queue-linked-list.cpp
// Content: Circular Queue using linked list
// Author: Tawfik Yasser
// Date: Mon, 8 Mar 2021
#include <iostream>
using namespace std;
class Node
{
public:
    int data;
    Node *next;
    Node() {}
};
class QueueLL
{
private:
    Node *front;
    Node *rear;
    int length;
public:
    QueueLL()
    {
        front = rear = NULL; length = 0;
    }
    //Function to add new element in the queue
    void enqueue(int value)
    {
        Node *newNode = new Node();
        newNode->data = value;
        newNode->next = NULL;
        if(rear == NULL)
        {
            front = rear = newNode;
            rear->next = front;
        }
        else
        {
            rear->next = newNode;
            rear = newNode;
            rear->next = front;
        }
      length++;
    }
    //Function to remove element from the queue
    void dequeue()
    {
        Node *temp = front;
        if(front == NULL && rear == NULL){
            cout<<"Empty Queue"<<endl;
        }else if(front == rear){
            front = rear = NULL;
            delete temp;
        }else{
            front = temp->next;
            rear->next = front;
            delete temp;
        }
      length--;
        if(length < 0 )
            length = 0;
    }
  //Function returns the first element in the queue
    void peek()
    {
        if(front == NULL && rear == NULL){
            cout<<"Empty Queue"<<endl;
        }else{
            cout<<front->data<<endl;
        }
    }
  //Function to display all the elements in the queue
    void print()
    {
        if(front == NULL && rear == NULL)
        {
            cout<<"Empty Queue"<<endl;
        }
        else
        {
            Node *temp = front;
            while(temp->next != front)
            {
                cout<<"[ "<<temp->data<<" ] ";
                temp = temp->next;
            }
            cout<<"[ "<<temp->data<<" ] ";
        }
    }
    int _size()
    {
        return length;
    }
};
int main()
{
    QueueLL qll;
    qll.enqueue(5);
    qll.enqueue(3);
    qll.enqueue(-2);
    qll.print();
    cout<<endl;
    cout<<"Queue size = "<<qll._size()<<endl;
    qll.peek();
    qll.dequeue();
    qll.print();
    cout<<endl;
    cout<<"Queue size = "<<qll._size()<<endl;
    qll.peek();
    return 0;
}
